README.md
# Fish 

![TravisCI build](https://travis-ci.org/nkreer/Fish.svg)
![License](https://img.shields.io/badge/License-GPL%20v3-blue.svg)

Fish is a **libre** and open source IRC-Bot with a powerful plugin API written in PHP. 
It is distributed under the terms and conditions of the [GPL Version 3](LICENSE).

## Getting started

### Requirements

* PHP 7 (Tested on 7.0.x, should work on 7.1 as well)
* Unix-like OS (Tested on macOS Sierra and Debian Linux)
* Composer

Fish has been tested to work well on unix systems and PHP 7.
It does not work with PHP 5 or Windows at the moment. HHVM hasn't been tested. 
To install and use Fish, make sure you also have [composer](https://getcomposer.org) installed. 

### Installation

For the best compatibility with plugins, you should always use [the latest stable release](https://github.com/nkreer/Fish/releases).
It's not recommended to use the source-code from this repository in production, as it may contain bugs or be unstable.
 
#### Normal installation

Either directly use composer to download and install the newest version:

`$ composer create-project nkreer/fish`
 
or download the latest codebase and let composer install the software using

`$ composer install`
 
#### Installation for use as a library

Fish is available on Packagist. You can require it for your own projects like this.

`$ composer require nkreer/fish`

Please keep in mind that the software isn't designed to be used as a library. Some features may not work in your environment. 

### Configuration

Configuration is done in the fish.json file. You can find help with setting Fish up in [CONFIGURATION.md](CONFIGURATION.md).

### Start

Open a terminal, navigate to the bot's source files and run

`$ php Start.php <address> [arguments]`

The available arguments are:

```
--port <port>           for using a port other than 6697

--no-ssl true           for connecting without TLS (Don't use this unless really needed)

--password <password>   for connecting to a passworded server

--config <path>         for using a different config file
```

## Features

Fish offers many features for interaction with an IRC Server out of the box: 

* Super simple and powerful plugin API (The most simple you will find in a PHP IRC-Bot to date!)
* Support for multiple connections in one process
* Built-in and adapting management-commands (join, part, help, etc.)
* Built-in user authentication features
* Built-in permission management
* Many more

## Extend

Fish can be extended by plugins. Several have already been written: 

| Plugin | Description |
| ------ | ----------- |
|[PluginTools](https://github.com/nkreer/PluginTools)| Helps you with  packaging of your plugins |
|[Permissions](https://github.com/nkreer/Permissions)| Changes users' permissions on IRC |
|[Scripts](https://github.com/nkreer/Fish-Scripts)| Enables you to add simple custom commands to the bot |

More is in the works.

## API Documentation

You can find a simple tutorial on how to write your own plugins in this repository's [wiki](https://github.com/nkreer/Fish/wiki).
If you have any questions regarding the API, feel free to send me an E-Mail.

CHANGELOG.md
# Changelog

Changelogs and release notes can also be found [on GitHub](https://github.com/nkreer/Fish/releases)!

### Releases

#### 1.2 Catfish

* Handle PRIVMSG ACTIONs the same as normal messages, include them in MessageReceiveEvent
* Fixes a crash with enabled user tracking on KickEvent

#### 1.1.1 Catfish

* Fixes a possible crash when requesting the user on ChannelLeaveEvent

#### 1.1 Catfsh

* Automatically rejoin a channel when kicked
* Automatically reconnect in case of a timeout
* Switch to an alternative nick when the primary is taken
* Support AWAY, LIST and INVITE
* Follow channel invitations from bot administrators
* reloadPlugin and uptime command
* Remove API versions (Plugins use the Fish version to tell for what version they have been developed)
* Fixed problems where the bot would not wait for the server to accept input

#### 1.0 Tuna (Initial Release)

* Super simple but powerful plugin API!
* Simple command framework!
* Command-spam protection!
* Simple event system!
* Task scheduler!
* Highly configurable!
* User authentication!
* Permission management!
* Multi-network support!
* Built in commands!

src/IRC/IRC.php
<?php

/*
 *
 * Fish - IRC Bot
 * Copyright (C) 2016 - 2017 Niklas Kreer
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

namespace IRC;

use IRC\Event\Connection\ConnectionActivityEvent;
use IRC\Event\Connection\ConnectionUnknownCommandEvent;
use IRC\Utils\BashColor;
use IRC\Utils\JsonConfig;

class IRC {

    const VERSION = "1.2";
    const CODENAME = "Catfish";

    const IRC_PORT = 6667;
    const IRC_PORT_ENCRYPTED = 6697;

    public static $instance;

    /**
     * @return IRC
     */
    public static function getInstance() : IRC{
        return self::$instance;
    }

    /**
     * @var Connection[]
     */
    public $connections = [];

    /**
     * @var JsonConfig
     */
    private $config;
    private $configPath;

    public $verbose = false;
    public $silent = false;

    public $idleTime = false;

    private $startupTime = 0;

    public function __construct(bool $verbose = false, bool $silent = false, String $configPath = "fish.json"){
        $this->silent = $silent;
        $this->verbose = $verbose;
        self::$instance = $this;
        $this->startupTime = time();
        Logger::info(BashColor::GREEN."Starting Fish (".self::CODENAME.") v".self::VERSION);
        @mkdir("plugins");
        @mkdir("users");
        $this->loadConfig($configPath);
    }

    public function reload(){
        $this->loadConfig($this->configPath);
        foreach($this->connections as $connection){
            $connection->getEventHandler()->unregisterAll();
            $connection->getScheduler()->cancelAll();
            $connection->load();
            $connection->getPluginManager()->reloadAll();
        }
    }

    public function stop(){
        Logger::info("Stopping...");
        foreach($this->connections as $connection){
            $this->removeConnection($connection);
            $connection->getPluginManager()->unloadAll();
        }
        exit(0);
    }

    public function loadConfig($path){
        if(!file_exists($path)){
            if($path === "fish.json"){
                Logger::info(BashColor::RED."Couldn't find configuration file. Making a new one...");
                @copy(__DIR__.DIRECTORY_SEPARATOR."Resources".DIRECTORY_SEPARATOR."fish.json", "fish.json");
            } else {
                Logger::info(BashColor::RED."No config found in ".$path.": Falling back to default...");
                $path = __DIR__.DIRECTORY_SEPARATOR."Resources".DIRECTORY_SEPARATOR."fish.json";
            }
        }
        $this->configPath = $path;
        $conf = new JsonConfig();
        $conf->loadFile($path);
        $this->config = $conf;
        $this->idleTime = $conf->getData("cpu_idle") * 1000;
    }

    public function getCommandPrefix() : String{
        return $this->getConfig()->getData("command_prefix")[0];
    }

    public function getConfig() : JsonConfig{
        return $this->config;
    }

    public function cycle(){
        foreach($this->connections as $connection){
            $new = $connection->check();
            if($new instanceof Command){
                $this->handle($new, $connection);
            }
            $this->runScheduler($connection);
        }
    }

    public function runScheduler(Connection $connection){
        //Do catchup calls if we've missed any
        if(time() - $connection->getScheduler()->getLastCall() >= 2){
            for($time = time(); $time > $connection->getScheduler()->getLastCall(); $time--){
                $connection->getScheduler()->call($time);
            }
        }

        //Regularly run scheduler
        $connection->getScheduler()->call();
    }

    public function handle(Command $run, Connection $connection){
        $ev = new ConnectionActivityEvent($connection, $run);
        $connection->getEventHandler()->callEvent($ev);
        if(!$ev->isCancelled()){
            $command = $run->getCommand();
            if(is_numeric($command)){
                $command = "_".$command;
            }
            $function = '\IRC\Protocol\\'.strtoupper($command);
            if(method_exists($function, "run")){
                // Most stupid hack ever, but it gets the job done
                call_user_func($function."::run", $run, $connection, $this->getConfig());
            } else {
                // Handle unknown commands
                $unknownEvent = new ConnectionUnknownCommandEvent($connection, $run);
                $connection->getEventHandler()->callEvent($unknownEvent);
            }
        }
    }

    /**
     * This keeps everything alive
     */
    public function run(){
        while(true){
            if(empty($this->connections)){
                $this->stop(); // Stop
            }
            $this->cycle();
            if($this->idleTime){
                usleep($this->idleTime); //Chill the CPU
            }
        }
    }

    /**
     * @param Connection $connection
     * @param bool $default
     * @return bool
     */
    public function addConnection(Connection $connection, bool $default = true) : bool{
        if(!$this->isConnected($connection->getAddress())){
            Logger::info(BashColor::CYAN."Connecting to ".$connection->getAddress().":".$connection->getPort()."...");
            //Setting up connection details
            if($default === true){
                $connection->nickname = $this->getConfig()->getData("default_nickname", "FishBot");
                $connection->realname = $this->getConfig()->getData("default_realname", "Fish - IRC Bot");
                $connection->username = $this->getConfig()->getData("default_username", "Fish");
                $connection->hostname = $this->getConfig()->getData("default_hostname", "Fish");
            }
            $result = $connection->connect();
            if($result){
                $this->connections[$connection->getAddress()] = $connection;
                Logger::info(BashColor::GREEN."Connected to ".$connection->getAddress().":".$connection->getPort());
                return true;
            } else {
                Logger::info(BashColor::RED."Can't connect to ".$connection->getAddress().":".$connection->getPort());
            }
        } else {
            Logger::info(BashColor::RED."Can't connect to ".$connection->getAddress().":".$connection->getPort().": Already connected");
        }
        return false;
    }

    /**
     * Close a connection
     * @param Connection $connection
     * @param String $quitMessage
     * @return bool
     */
    public function removeConnection(Connection $connection, String $quitMessage = null) : bool{
        if($this->isConnected($connection->getAddress())){
            Logger::info(BashColor::RED."Disconnecting ".$connection->getAddress().":".$connection->getPort());
            if($quitMessage === null){
                $quitMessage = $this->config->getData("default_quitmsg", "Leaving");
            }
            $connection->disconnect($quitMessage);
            unset($this->connections[$connection->getAddress()]);
            return true;
        } else {
            unset($this->connections[$connection->getAddress()]);
            return true; //These statements are kept for reasons of backwards-compatibility
        }
    }

    /**
     * @param $address
     * @return bool
     */
    public function isConnected(String $address) : bool{
        return isset($this->connections[$address]);
    }

    public function getStartupTime(){
        return $this->startupTime;
    }

}